### ISSUE TITLE
Nested schema is not being instantiated when it's a `Union` that uses `ForwardRef`

### ISSUE THREAD

**Issue by Michael-J-Ward** [samuelcolvin/pydantic#404]
# Bug
Nested schema is not being instantiated when it's a `Union` that uses `ForwardRef`

For bugs/questions:
* OS: cat /etc/os-release
```
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```
* Python version `import sys; print(sys.version)`: **'3.7.2 (default, Feb  6 2019, 12:04:03) \n[GCC 6.3.0 20170516]'**
* Pydantic version `import pydantic; print(pydantic.VERSION)`: **StrictVersion ('0.20')**

Where possible please include a self contained code snippet describing your 
bug, question, or where applicable feature request:

```py
from pydantic import BaseModel
from typing import Union, ForwardRef

class Leaf(BaseModel):
    a: str
    b: int


TreeType = Union[ForwardRef('Node'), Leaf]


class Node(BaseModel):
    value: int
    left: TreeType
    right: TreeType


data = {
    "value": 3,
    "left": {
        "a": "foo",
        "b": 2,
    },
    "right": {
        "value": 5,
        "left": {
            "a": "bar",
            "b": 1,
        },
        "right": {
            "a": "buzz",
            "b": 7
        }  
    }
}

node = Node(**data)
assert isinstance(node.left, Leaf), str(type(node.left))
assert isinstance(node.right, Node), str(type(node.right))
```

```
Traceback (most recent call last):
  File "foo.py", line 38, in <module>
    assert isinstance(node.left, Leaf), str(type(node.left))
AssertionError: <class 'dict'>
```

**Comment by samuelcolvin** [comment]
@tiangolo any idea?

My suspicion is that this is currently beyond what's possible, but it might possible to somehow call `update_forward_refs()` before generating the schema.

**Comment by tiangolo** [comment]
I haven't been able to see the new ForwardRef features yet, but I'll check it as soon as I can.

It will take me a bit as I'll be mostly offline on a trip this week.

**Comment by mooncake4132** [comment]
I'm also experiencing this issue when using creating the nested schemas with `Optional` as `Optional`s get translated into `Union`s. Maybe [update_forward_refs](https://github.com/samuelcolvin/pydantic/blob/a704662ae4cde638c0e1b68126a2f89828787641/pydantic/main.py#L456) can be updated to recursively traverse `__args__` for `Union`s?

**Comment by tiangolo** [comment]
This is (potentially) addressed in #464.

Have in mind that apart from the PR, for your example to work, you need to call `Node.update_forward_refs()` in your code right after creating the `Node` class.

The change in the PR makes `update_forward_refs()` recursively update forward refs of sub-fields, that includes `Union`, `Tuple`, etc.

The PR includes a test with the same example you have here (in a slimmed down version).
